.github/actions/do_build_dpcpp/action.yml

```diff
@@ -0,0 +1,122 @@
+name: build dpc++
+description: build dpc++
+
+inputs:
+  target:
+    description: 'target architecture'
+  download_dpcpp_artefact:
+    description: 'none | dpcpp_release' # TODO: add support for 'previous workflow' download
+    type: string
+    default: "none"
+
+runs:
+  using: "composite"
+  steps:
+    - name: calc vars
+      id: calc_vars
+      uses: ./.github/actions/calc_vars
+      with:
+        target: ${{ inputs.target }}  
+
+    - name: Install Ninja
+      uses: llvm/actions/install-ninja@main
+
+    - name: Install spirv tools
+      shell: bash
+      run:    
+        sudo apt-get install spirv-tools
+    
+    - name: clone dpc++
+      if: inputs.download_dpcpp_artefact == 'none'
+      uses: actions/checkout@v4
+      with:
+        repository: intel/llvm
+        path: llvm
+
+    - name: dpcpp configure
+      if: inputs.download_dpcpp_artefact == 'none'
+      shell: bash
+      run:
+        cd llvm; python buildbot/configure.py
+        -o build
+        --host-target="X86;AArch64;RISCV"
+        --native_cpu
+        --llvm-external-projects=lld
+        --cmake-opt=-DNATIVECPU_USE_OCK=ON
+        --cmake-opt=-DLLVM_ENABLE_ZLIB=OFF
+        --cmake-opt=-DLLVM_ENABLE_ZSTD=OFF
+
+    - name: build sycl-headers
+      if: inputs.download_dpcpp_artefact == 'none'
+      shell: bash
+      run:
+        cmake --build llvm/build -- sycl-headers
+  
+    - name: build dpc plus plus
+      if: inputs.download_dpcpp_artefact == 'none'
+      shell: bash
+      run:
+       python llvm/buildbot/compile.py -o llvm/build -v -j 8
+
+    - name: build extra utilties
+      if: inputs.download_dpcpp_artefact == 'none'
+      # Build various utilities, since those aren't proper dependencies.
+      # FileCheck and not are needed for tests. The rest are needed for
+      # cross builds. They are enabled on all targets for consistency.
+      shell: bash
+      run:
+       cmake --build llvm/build -- FileCheck clang-tblgen
+        libclc-remangler llvm-as llvm-min-tblgen llvm-tblgen not
+        opt prepare_builtins -j8
+
+    - name: copy utilities
+      if: inputs.download_dpcpp_artefact == 'none'
+      shell: bash
+      run:
+        cd llvm/build/bin;
+        cp FileCheck clang-tblgen libclc-remangler llvm-as llvm-min-tblgen
+        llvm-tblgen not opt prepare_builtins ../install/bin
+
+    # TODO: For now just linux x86_64
```

Comment:
This could go in if you apply the patches. However could be delayed for a further PR.

Reply:
I'll defer until the SYCL-CTS aarch64/riscv ticket (which is likely next on the list in any case)